server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+          length = 0;
+          s3_log(S3_LOG_INFO, stripped_request_id,
+                 "No further prefix match. Skipping further object versions "
+                 "listing\n");
+          break;
+        }
+      }
+    }
+
+    if ((--length == 0) ||
+        (!last_key_in_common_prefix &&
+         ((versions_list.size() + common_prefixes.size()) == max_keys))) {
+      // This is the last element returned or we reached limit requested, we
+      // break.
+      last_key = kv.first;
+      break_out = true;
+    }
+  }  // End of for loop
+
+  if (json_error) {
+    s3_iem(LOG_ERR, S3_IEM_METADATA_CORRUPTED, S3_IEM_METADATA_CORRUPTED_STR,
+           S3_IEM_METADATA_CORRUPTED_JSON);
+  }
+
+  // We ask for more if there is any.
+  key_Count = versions_list.size() + common_prefixes.size();
+  if ((key_Count == max_keys) ||
+      (!skip_remaining_common_prefixes && (kvps.size() < max_record_count)) ||
+      (no_further_prefix_match)) {
+    // Go ahead and respond.
+    if (key_Count == max_keys && length != 0) {
+      // When we hit the max keys condition and previously we skipped common
+      // prefix keys (i.e. skip_remaining_common_prefixes = true),
+      // we can't rely on 'length' variable for truncation flag.
+      // In such case, we don't know if there are any more keys left after
+      // skipping the common prefix. In such situation, we have to explictly
+      // identify if there are any further keys left in a bucket to mark
+      // the truncation flag in the response.
+      if (skip_remaining_common_prefixes) {
+        // Check with Motr if there are any keys to be seen.
+        // If there are still some keys, we would set truncation flag, else no.
+        check_any_keys_after_prefix = true;
+        if (last_key_in_common_prefix) {
+          saved_last_key = last_common_prefix;
+        }
+        // Call get_next_versions to see remaining keys, if any, after
+        // skipping keys belonging to common prefix.
+        get_next_versions();
+        return;
+      } else {
+        response_is_truncated = true;
+        // Before sending response, check if the previous key was in common
+        // prefix. If yes, we need to return the common prefix as next marker.
+        if (last_key_in_common_prefix) {
+          last_key = last_common_prefix;
+          add_marker_version(last_common_prefix, "");
+        } else {
+          auto& version = versions_list.back();
+          add_marker_version(version->get_object_name(),
+                             version->get_obj_version_id());
+        }
+        s3_log(S3_LOG_DEBUG, request_id, "Next marker = %s\n",
+               last_key.c_str());
+      }
+    }
+    fetch_successful = true;
+    next();
+  } else {
+    get_next_versions();
+  }
+}
+
+void S3ListObjectVersionsAction::get_next_versions_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (motr_kv_reader->get_state() == S3MotrKVSReaderOpState::missing) {
```

Comment:
Personal preference, I think a `switch` here is easier to read.

Reply:
Done